### ISSUE TITLE
SSSD as a sidecar container causing issues

### ISSUE THREAD

**Issue by brianmajor** [opencadc/science-platform#44]
Bundling the SSSD container in the science container pods is causing three notable issues:

1.  Often the science container completes initialization before the SSSD container.  This causes the bash shell to not pickup the current user correctly.
2.  When the science container exits, the pod does not terminate because the SSSD container is still running.
3.  SSSD currently needs to be run as root.  Because the granularity of privileged access is at the pod level, the entire pod (including the science container) must be run with privileged access.  This will probably not be acceptable to infrastructure providers.

**Comment by rptaylor** [comment]
Re 3: 
It is possible (according to the docs) to have just one container in the pod be privileged:
https://kubernetes.io/docs/tasks/configure-pod-container/security-context/#set-the-security-context-for-a-container

However,  RBAC and PSP do not provide any mechanism for _enforcing_ security contexts with container-level (sub-pod) granularity, so we would have to allow *all* user pods and containers the ability to run privileged. This means we can not provide you any protection; if there is a bug or exploit in ARCADE that accidentally configures the wrong container to be privileged, users will have root access.  Aside from that, I am not sure if it is possible to break into one container from another in the same pod (there is definitely less segregation between containers than between pods). but this would allow privilege escalation.
Overall I am keeping in mind that it is relatively common for end user passwords/credentials to be compromised.

**Comment by rptaylor** [comment]
The alternative approach that we mulled over awhile ago is: 
A configmap or secret can be mounted into a pod as a file containing arbitrary text. The user identity information could be stored in a configmap/secret in the format of passwd, group, etc files and mounted into the pods as /etc/passwd /etc/group etc.  When the configmap or secret is updated, those updates are propagated to all pods live (may take a few seconds) so all nodes can see the updated user info. 

This has the advantage of fully leveraging native k8s functionality, reducing the security liability of requiring a privileged container to be running in every user pod (and slight resource usage reduction for the SSSD container). Disadvantage is requiring some extra scriptage to extract the identity info from LDAP and construct passwd and group files (should be easy with Perl-LDAP or similar). 

I started testing this out and came up with a PoC that initially seemed very promising. However, I ran into an edge case:  secrets and configmaps only get updated live on nodes if they are mounted as whole directories, without the subpath volume option. On the other hand, we require the subpath option in order for the configmap/secret to be mounted as just an individual file.

There is a little note here describing this gotcha:
https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/#add-configmap-data-to-a-specific-path-in-the-volume
"Note: A container using a ConfigMap as a subPath volume will not receive ConfigMap updates."
and this corresponding issue:
https://github.com/kubernetes/kubernetes/issues/50345

It is still possible to work around this limitation using symlinks and scriptage that runs on pod startup, as described here: 
https://github.com/kubernetes/kubernetes/issues/50345#issuecomment-391888999
https://github.com/kubernetes/kubernetes/issues/50345#issuecomment-400647420
This would work but may not be robust for processes running in the pod, because it means the passwd files would change while the pod is running, rather than being fully set up as desired before the pod starts.
